src/binary-encoding.adoc

```diff
@@ -0,0 +1,1600 @@
+│┌─── Low nibble 0 indicates
+││    no more bytes follow.
+50
+----
+
+==== Example encoding of `17`
+[source]
+----
+┌──── Opcode in 50-58 range indicates integer
+│┌─── Low nibble 1 indicates
+││    a single byte follows.
+51 11
+    └── FixedInt 17
+----
+
+==== Example encoding of `-944`
+[source]
+----
+┌──── Opcode in 50-58 range indicates integer
+│┌─── Low nibble 2 indicates
+││    that two bytes follow.
+52 50 fc
+   └─┬─┘
+FixedInt -944
+----
+
+==== Example variable-length encoding of `-944`
+[source]
+----
+┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
+│   ┌─── VarUInt 2; a 2-byte FixedInt follows
+│   │    that two bytes follow.
+F5 05 50 fc
+      └─┬─┘
+   FixedInt -944
+----
+
+[[floats]]
+=== Floats
+
+Float values are encoded using the IEEE-754 specification, and can be serialized in four sizes:
+
+* 0 bits (0 bytes), representing the value 0e0 and indicated by opcode `0x5A`
+* 16 bits (2 bytes, link:https://en.wikipedia.org/wiki/Half-precision_floating-point_format[half precision]),
+indicated by opcode `0x5B`
+* 32 bits (4 bytes, link:https://en.wikipedia.org/wiki/Single-precision_floating-point_format[single precision]),
+indicated by opcode `0x5C`
+* 64 bits (8 bytes, link:https://en.wikipedia.org/wiki/Double-precision_floating-point_format[double precision]),
+indicated by opcode `0x5D`
+
+Note that in the Ion data model, float values are always 64 bits. However, if a value can be losslessly serialized
+in fewer than 64 bits, applications may choose to do so.
+
+==== Example encoding of `0e0`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble A indicates
+││    a 0-length float; 0e0
+5A
+----
+
+==== Example encoding of `3.14`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble B indicates a 2-byte float
+││
+5B 42 47
+   └─┬─┘
+half-precision 3.14
+----
+
+==== Example encoding of `3.1415927`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble C indicates a 4-byte,
+││    single-precision value.
+5C 40 49 0F DB
+   └────┬────┘
+single-precision 3.1415927
+----
+
+==== Example encoding of `3.141592653589793`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble C indicates a 4-byte,
+││    single-precision value.
+5D 40 09 21 FB 54 44 2D 18
+   └──────────┬──────────┘
+double-precision 3.141592653589793
+----
+
+[[decimals]]
+=== Decimals
+
+If an opcode has a high nibble of `0x6_`, it represents a decimal. Low nibble values `0x_E` and below indicate
+the number of trailing bytes used to encode the decimal.
+
+The body of the decimal is encoded as a <<varint, `VarInt`>> representing its exponent, followed by a `FixedInt`
+representing its coefficient. The width of the coefficient is the total length of the decimal encoding minus the length
+of the exponent. It is possible for the coefficient to have a width of zero, indicating a coefficient of `0`.
```

Comment:
> Wow, this is silly! It would make more sense for the exponent to be optional, to shave a byte from decimal-encoded integers like 17. aka 17d0. Eliding the coefficient shaves bits for the far less useful set of values 0.000....

Wow, that's a great point! I ported over the 1.0 representation without too much thought, but your approach sounds much more beneficial. I'll take a pass at writing this up.